---
title: "TCGA-DE"
date: "4/17/2019"
output: html_document
---

Conduct differential expression analysis using limma-voom pipeline on TCGA HCC dataset, follow the [RNASeq 123 workflow](https://bioconductor.org/packages/devel/workflows/vignettes/RNAseq123/inst/doc/limmaWorkflow.html#organising-gene-annotations):

- Compare between `hcc tumoral tissues` and `paired adjacent normal tissues` 
- The count data we have are gene-level HTSeq counts, which generated by `HTSeq`

```{R set up}
suppressPackageStartupMessages({
  library(limma)
  library(edgeR)
  library(DESeq2)
  library(dplyr)
  library(pheatmap)
})
source('~/hcc/scripts/functions.R')
# Load data
data.dir <- '~/hcc/data/TCGA'
load(file.path(data.dir, "hcc_TCGA.Rdata"))
```

## Preprocess data
### Get rid of the recurrent samples
```{R}
sheet <- sheet[sheet$Sample.Type != 'Recurrent Tumor', ]
colnames(sheet)[6] <- 'submitter_id'
samples <- merge(sheet[, c(6,7,8)], clinic, by='submitter_id', all=T)

filt.counts <- data.matrix(filt.counts) # df -> numeric matrix
colnames(filt.counts) <- gsub('\\.', '-', colnames(filt.counts))
filt.counts <- filt.counts[, colnames(filt.counts) %in% samples$Sample.ID]
dim(filt.counts)

samples <- samples[match(colnames(filt.counts), samples$Sample.ID),]
# Temporarily remove that sample with missing age 
rid.id <- samples[which(is.na(samples$age_at_initial_pathologic_diagnosis)),]$Sample.ID
samples <- samples[samples$Sample.ID != rid.id, ]
filt.counts <- filt.counts[, colnames(filt.counts) != rid.id]
all(colnames(filt.counts) == samples$Sample.ID)
```


### Filter out lowly-expressed genes 
Count matrix with rows of genes and columns of samples
```{R, fig.width=9}
dge <- DGEList(counts=filt.counts)
dge$samples$group <- factor(samples$Sample.Type)
table(rowSums(dge$counts==0)==ncol(dge)) # zero counts across all nine samples
n.lcpm <- cpm(dge, log=T)
# remove rows that consistently have zero or very low counts
keep <- filterByExpr(dge, group=dge$samples$group)
dge <- dge[keep, keep.lib.sizes=F]
cpm <- cpm(dge)
lcpm <- cpm(dge, log=T)

col <- getColors(ncol(lcpm))
par(mfrow=c(1 ,2))
dens.plot(n.lcpm, col, c(0,0.6))
dens.plot(lcpm, col, c(0,0.6))
```


## TMM normalization
```{R norm, fig.width=9}
dge <- calcNormFactors(dge, method="TMM")
dge$samples$norm.factors[1:5]

x2 <- dge[, c(1:10)]
x2$samples$norm.factors <- 1
x2$counts[,1] <- ceiling(x2$counts[,1]*0.05)
x2$counts[,2] <- x2$counts[,2]*5

par(mfrow=c(1,2))
lcpm <- cpm(x2, log=TRUE)
boxplot(lcpm, las=2, col=col, main="")
title(main="A. Example: Unnormalised data",ylab="Log-cpm")
x2 <- calcNormFactors(x2)  
x2$samples$norm.factors

lcpm <- cpm(x2, log=TRUE)
boxplot(lcpm, las=2, col=col, main="")
title(main="B. Example: Normalised data",ylab="Log-cpm")
```

### Clustering of samples
```{R cluster}
lcpm <- cpm(dge, log=T)

col.group <- samples$Sample.Type
levels(col.group) <-  brewer.pal(nlevels(col.group), "Set1")
col.group <- as.character(col.group)
plotMDS(lcpm, labels=NULL, pch=".", cex = 2.5, col=col.group)
title(main="Sample groups")
```


## Differential expression analysis
### Create a design matrix 
```{R design}
tissue <- factor(ifelse(samples$Sample.Type == 'Primary Tumor', 1, 0))
sex <- samples$gender
race <- samples$race
age <- samples$age_at_initial_pathologic_diagnosis
#stage <- samples$tumor_stage

tumor <- 'tissue1'
design <- model.matrix(~ tissue + sex + race + age)
head(design, 5)
```


### Removing heteroscedascity from count data
When operating on a DGEList-object, voom converts raw counts to log-CPM values by automatically extracting library sizes and normalisation factors from x itself. 

```{R}
v <- voom(dge, design, plot=T)
vfit <- lmFit(v, design)
efit <- eBayes(vfit)
plotSA(efit, main="Final model: Mean-variance trend")
```


### Generate top table
```{R tt}
tt <- topTable(efit, number=Inf, coef=tumor)
head(tt, 15)
limma::volcanoplot(efit, coef = tumor, highlight = 3, names = rownames(efit))
limma::plotMA(efit, main = "TCGA HCC")
# why the logFC in MA plot are so small????
```


### Significant gene numbers
```{R}
nrow(tt[tt$adj.P.Val < 0.01, ])
nrow(tt[(tt$adj.P.Val < 0.01) & (abs(tt$logFC) > log2(2)), ])
nrow(tt[(tt$adj.P.Val < 0.01) & (abs(tt$logFC) > log2(3)), ])
```


### Heatmap
```{R heatmap, fig.width=12, fig.height=10}
draw_heatmap <- function(voomObj, topTable, phenoDF, list) {
  hm_cdr <- phenoDF %>% select(Sample.Type, tumor_stage)
  rownames(hm_cdr) <- colnames(voomObj)
  colnames(hm_cdr)  <- c('tumor', 'stage')
  tumor <- c("#99a599", "#37637f")
  names(tumor) <- unique(hm_cdr$tumor)
  stage <- c("#d7191c","#fdae61","#a1d99b","#2b83ba","#bababa")
  names(stage) <- c("I","II","III","IV","NA")
  anno_colors <- list(tumor = tumor,
                      stage = stage) # the name must be consistent
  h <- pheatmap(voomObj$E[list, ], annotation_col=hm_cdr, annotation_colors=anno_colors, 
                labels_row = list, show_colnames = F)
  h
}

draw_heatmap(v, tt, samples, rownames(tt)[1:50])
```


### Jitter plot
```{R jittor, fig.height=7}
for (id in c('CLEC4M', 'CLEC4G', 'CLEC1B')) {
  print(single.box(v, samples, id, tt))
}
```


### Save 
```{R save}
save(samples, dge, tt, file=file.path(data.dir, 'DE.TCGA.Rdata'))
```